### PR TITLE
Interception tests

### LINKED ISSUE #19: Add interception.spec.js tests
Based on this tag https://github.com/kblok/playwright/tree/first-snapshot. We don't need to get the tests green. We only need to incoporate the code. Let's also check if we can get code from Puppeteer

### PR DESCRIPTION
**Pull request by Meir017**
closes #19

### CODE REVIEW

**Comment by kblok** [comment]
@Meir017 you should be able to call "dotnet format" locally.
On the other hand, I changed line endings to CRLF, which should make your life easier.

**Comment by kblok** [comment]
I just confirmed that your files are changing CRLF back to LF, maybe your IDE didn't get the editorconfig changes.

**Comment by Meir017** [comment]
@kblok hmm, I used `dotnet-format` locally. does it need to be configured?

**Comment by kblok** [comment]
Use this version
```
dotnet tool install -g dotnet-format --add-source https://dotnet.myget.org/F/format/api/v3/index.json
```

and then run `dotnet format` but make sure you are on the `src` folder.

**Comment by Meir017** [comment]
@kblok I installed dotnet through wsl and then the formatting worked 😄

File: src/PlaywrightSharp.Abstractions/Extensions.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    public static class Extensions
+    {
+        /// <summary>
+        /// Runs <paramref name="pageFunction"/> within the frame and passes it the outcome of <paramref name="elementHandleTask"/> as the first argument.
+        /// </summary>
+        /// <param name="elementHandleTask">A task that returns an <see cref="IElementHandle"/> that will be used as the first argument in <paramref name="pageFunction"/>.</param>
+        /// <param name="pageFunction">Function to be evaluated in browser context.</param>
+        /// <param name="args">Arguments to pass to <c>pageFunction</c>.</param>
+        /// <returns>Task.</returns>
+        public static async Task EvaluateAsync(this Task<IElementHandle> elementHandleTask, string pageFunction, params object[] args)
```
Comment (kblok) [inline src/PlaywrightSharp.Abstractions/Extensions.cs]: We are going to get closer to the Playwright API here. We already have a `QuerySelectorEvaluateAsync` method on IPage and IFrame.

File: src/PlaywrightSharp.Abstractions/IPage.cs [hunk 2]
```diff
@@ -242,6 +242,22 @@ public interface IPage
...
+        /// </summary>
+        /// <example>
+        /// <code>
+        /// <![CDATA[
+        /// var firstRequest = await page.WaitForRequestAsync("http://example.com/resource");
+        /// return firstRequest.Url;
+        /// ]]>
+        /// </code>
+        /// </example>
+        /// <returns>A task which resolves when a matching request was made.</returns>
```
Comment (kblok) [inline src/PlaywrightSharp.Abstractions/IPage.cs]:
```suggestion
        /// <returns>A <see cref="Task"/> that completes when the request was made (or timeout), yielding the matching <see cref="IRequest">.</returns>

```

File: src/PlaywrightSharp.Abstractions/IRequest.cs [hunk 3]
```diff
@@ -11,9 +15,98 @@ public interface IRequest
...
+
+        /// <summary>
+        /// Gets or sets the HTTP headers.
+        /// </summary>
+        IDictionary<string, string> Headers { get; }
+
+        /// <summary>
+        /// Gets or sets the post data.
+        /// </summary>
+        object PostData { get; }
```
Comment (kblok) [inline src/PlaywrightSharp.Abstractions/IRequest.cs]:
```suggestion
        string PostData { get; }

```

File: src/PlaywrightSharp.Abstractions/IRequest.cs [hunk 4]
```diff
@@ -11,9 +15,98 @@ public interface IRequest
...
+        /// </code>
+        /// If the website https://google.com has no redirects, then the chain will be empty:
+        /// <code>
+        /// var response = await page.GoToAsync("https://google.com");
+        /// var chain = response.Request.RedirectChain;
+        /// Console.WriteLine(chain.Length); // 0
+        /// </code>
+        /// </example>
+        /// <value>The redirect chain.</value>
+        public IRequest[] RedirectChain { get; }
```
Comment (kblok) [inline src/PlaywrightSharp.Abstractions/IRequest.cs]:
```suggestion
        IRequest[] RedirectChain { get; }
```

File: src/PlaywrightSharp.Abstractions/IRequest.cs [hunk 5]
```diff
@@ -11,9 +15,98 @@ public interface IRequest
...
+
+        /// <summary>
+        /// Responsed attached to the request.
+        /// </summary>
+        public IResponse Response { get; internal set; }
```
Comment (kblok) [inline src/PlaywrightSharp.Abstractions/IRequest.cs]:
```suggestion
       IResponse Response { get; }

```

File: src/PlaywrightSharp.Abstractions/IRequest.cs [hunk 6]
```diff
@@ -11,9 +15,98 @@ public interface IRequest
...
+        /// Gets or sets the failure.
+        /// </summary>
+        string Failure { get; }
+
+        /// <summary>
+        /// Continues request with optional request overrides. To use this, request interception should be enabled with <see cref="IPage.SetRequestInterceptionAsync(bool)"/>. Exception is immediately thrown if the request interception is not enabled.
+        /// If the URL is set it won't perform a redirect. The request will be silently forwarded to the new url. For example, the address bar will show the original url.
+        /// </summary>
+        /// <param name="overrides">Optional request overwrites.</param>
+        /// <returns>Task.</returns>
```
Comment (kblok) [inline src/PlaywrightSharp.Abstractions/IRequest.cs]:
```suggestion
        /// <returns>A <see cref="Task"/> that completes when the message was confirmed by the browser.</returns>
```

File: src/PlaywrightSharp.Abstractions/IRequest.cs [hunk 7]
```diff
@@ -11,9 +15,98 @@ public interface IRequest
...
+        Task ContinueAsync(Payload overrides = null);
+
+        /// <summary>
+        /// Fulfills request with given response. To use this, request interception should be enabled with <see cref="IPage.SetRequestInterceptionAsync(bool)"/>. Exception is thrown if request interception is not enabled.
+        /// </summary>
+        /// <param name="response">Response that will fulfill this request.</param>
+        /// <returns>Task.</returns>
```
Comment (kblok) [inline src/PlaywrightSharp.Abstractions/IRequest.cs]:
```suggestion
        /// <returns>A <see cref="Task"/> that completes when the message was confirmed by the browser.</returns>
```

File: src/PlaywrightSharp.Abstractions/IRequest.cs [hunk 8]
```diff
@@ -11,9 +15,98 @@ public interface IRequest
...
+        Task RespondAsync(ResponseData response);
+
+        /// <summary>
+        /// Fulfills request with given response. To use this, request interception should be enabled with <see cref="IPage.SetRequestInterceptionAsync(bool)"/>. Exception is thrown if request interception is not enabled.
+        /// </summary>
+        /// <param name="response">Response that will fulfill this request.</param>
+        /// <returns>Task.</returns>
```
Comment (kblok) [inline src/PlaywrightSharp.Abstractions/IRequest.cs]:
```suggestion
        /// <returns>A <see cref="Task"/> that completes when the message was confirmed by the browser.</returns>
```

File: src/PlaywrightSharp.Abstractions/IRequest.cs [hunk 9]
```diff
@@ -11,9 +15,98 @@ public interface IRequest
...
+        Task FulfillAsync(ResponseData response);
+
+        /// <summary>
+        /// Aborts request. To use this, request interception should be enabled with <see cref="IPage.SetRequestInterceptionAsync(bool)"/>.
+        /// Exception is immediately thrown if the request interception is not enabled.
+        /// </summary>
+        /// <param name="errorCode">Optional error code. Defaults to <see cref="RequestAbortErrorCode.Failed"/>.</param>
+        /// <returns>Task.</returns>
```
Comment (kblok) [inline src/PlaywrightSharp.Abstractions/IRequest.cs]:
```suggestion
        /// <returns>A <see cref="Task"/> that completes when the message was confirmed by the browser.</returns>
```

File: src/PlaywrightSharp.Abstractions/Payload.cs [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+{
+    /// <summary>
+    /// Payload information.
+    /// </summary>
+    public class Payload
+    {
+        /// <summary>
+        /// Gets or sets the HTTP method.
+        /// </summary>
+        /// <value>HTTP method.</value>
```
Comment (kblok) [inline src/PlaywrightSharp.Abstractions/Payload.cs]: Remove value attribute

File: src/PlaywrightSharp.Abstractions/Payload.cs [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+
+        /// <summary>
+        /// Gets or sets the post data.
+        /// </summary>
+        /// <value>The post data.</value>
+        public string PostData { get; set; }
+
+        /// <summary>
+        /// Gets or sets the HTTP headers.
+        /// </summary>
```
Comment (kblok) [inline src/PlaywrightSharp.Abstractions/Payload.cs]: Remove value attribute

File: src/PlaywrightSharp.Abstractions/Payload.cs [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+        /// <value>HTTP headers.</value>
+        public IDictionary<string, string> Headers { get; set; } = new Dictionary<string, string>();
+
+        /// <summary>
+        /// Gets or sets the URL.
+        /// </summary>
+        /// <value>The URL.</value>
```
Comment (kblok) [inline src/PlaywrightSharp.Abstractions/Payload.cs]: Remove value attribute

File: src/PlaywrightSharp.Abstractions/Extensions.cs (hunk 1 above)
Comment (kblok) [reply src/PlaywrightSharp.Abstractions/Extensions.cs]: Let's move back the SelectionException to the PlaywrightSharp library.

File: src/PlaywrightSharp.Tests/RequestInterception/PageSetRequestInterceptionTests.cs [hunk 13]
```diff
@@ -0,0 +1,619 @@
...
+            Assert.True(response.Ok);
+        }
+
+        ///<playwright-file>interception.spec.js</playwright-file>
+        ///<playwright-describe>Page.setRequestInterception</playwright-describe>
+        ///<playwright-it>should work when POST is redirected with 302</playwright-it>
+        [Fact]
+        public async Task ShouldWorkWhenPOSTIsRedirectedWith302()
+        {
+
```
Comment (kblok) [inline src/PlaywrightSharp.Tests/RequestInterception/PageSetRequestInterceptionTests.cs]: remove empty line. Let's review this across this file.

File: src/PlaywrightSharp.Tests/RequestInterception/PageSetRequestInterceptionTests.cs [hunk 14]
```diff
@@ -0,0 +1,619 @@
...
+                var headers = new Dictionary<string, string>(e.Request.Headers) { ["foo"] = "bar" };
+                headers.Remove("origin");
+                e.Request.ContinueAsync(new Payload { Headers = headers });
+            };
+
+            var originRequestHeader = Server.WaitForRequest("/empty.html", request => request.Headers["origin"]);
+            await Task.WhenAll(
+             originRequestHeader,
+             Page.GoToAsync(TestConstants.EmptyPage)
+         );
```
Comment (kblok) [inline src/PlaywrightSharp.Tests/RequestInterception/PageSetRequestInterceptionTests.cs]: check indent

File: src/PlaywrightSharp.Tests/RequestInterception/PageSetRequestInterceptionTests.cs [hunk 15]
```diff
@@ -0,0 +1,619 @@
...
+            await Page.GoToAsync(TestConstants.EmptyPage);
+            Server.SetRedirect("/logo.png", "/pptr.png");
+            await Page.SetRequestInterceptionAsync(true);
+            Page.Request += async (sender, e) => await e.Request.ContinueAsync();
+            var status = await Page.EvaluateAsync<int>(@"async () => {
+            var request = new XMLHttpRequest();
+            request.open('GET', '/logo.png', false);  // `false` makes the request synchronous
+            request.send(null);
+            return request.status;
+        }");
```
Comment (kblok) [inline src/PlaywrightSharp.Tests/RequestInterception/PageSetRequestInterceptionTests.cs]: check indent

File: src/PlaywrightSharp.Tests/RequestInterception/PageSetRequestInterceptionTests.cs [hunk 16]
```diff
@@ -0,0 +1,619 @@
...
+        ///<playwright-file>interception.spec.js</playwright-file>
+        ///<playwright-describe>Page.setRequestInterception</playwright-describe>
+        ///<playwright-it>should be abortable</playwright-it>
+        [Fact]
+        public async Task ShouldBeAbortable()
+        {
+            await Page.SetRequestInterceptionAsync(true);
+            Page.Request += async (sender, e) =>
+            {
+                if (e.Request.Url.EndsWith(".css"))
```
Comment (kblok) [inline src/PlaywrightSharp.Tests/RequestInterception/PageSetRequestInterceptionTests.cs]: braces

File: src/PlaywrightSharp.Tests/RequestInterception/PageSetRequestInterceptionTests.cs [hunk 17]
```diff
@@ -0,0 +1,619 @@
...
+            await Page.SetRequestInterceptionAsync(true);
+            Page.Request += (sender, e) =>
+            {
+                e.Request.AbortAsync(RequestAbortErrorCode.InternetDisconnected);
+            };
+            IRequest failedRequest = null;
+            Page.RequestFailed += (sender, e) => failedRequest = e.Request;
+            await Page.GoToAsync(TestConstants.EmptyPage).ContinueWith(task => { });
+            Assert.NotNull(failedRequest);
+            if (TestConstants.IsWebKit)
```
Comment (kblok) [inline src/PlaywrightSharp.Tests/RequestInterception/PageSetRequestInterceptionTests.cs]: braces here

**Comment by kblok** [comment]
We should ignore the testCert.cer file

File: src/PlaywrightSharp.Tests/RequestInterception/PageAuthenticateTests.cs [hunk 18]
```diff
@@ -0,0 +1,73 @@
+using System.Net;
+using System.Threading.Tasks;
+using PlaywrightSharp.Tests.BaseTests;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace PlaywrightSharp.Tests.RequestInterception
```
Comment (kblok) [inline src/PlaywrightSharp.Tests/RequestInterception/PageAuthenticateTests.cs]: Let's put this file in the Page folder

File: src/PlaywrightSharp.Tests/RequestInterception/InterceptionSetOfflineModeTests.cs [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+    public class InterceptionSetOfflineModeTests : PlaywrightSharpPageBaseTest
+    {
+        internal InterceptionSetOfflineModeTests(ITestOutputHelper output) : base(output)
+        {
+        }
+
+        ///<playwright-file>interception.spec.js</playwright-file>
+        ///<playwright-describe>Interception.setOfflineMode</playwright-describe>
+        ///<playwright-it>should work</playwright-it>
+        [Fact]
```
Comment (kblok) [inline src/PlaywrightSharp.Tests/RequestInterception/InterceptionSetOfflineModeTests.cs]: Skip firefox

File: src/PlaywrightSharp.Tests/RequestInterception/InterceptionSetOfflineModeTests.cs [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+            var exception = Assert.ThrowsAsync<PlaywrightSharpException>(() => Page.GoToAsync(TestConstants.EmptyPage));
+            await Page.SetOfflineModeAsync(false);
+            var response = await Page.GoToAsync(TestConstants.EmptyPage);
+            Assert.Equal(HttpStatusCode.OK, response.Status);
+        }
+
+        ///<playwright-file>interception.spec.js</playwright-file>
+        ///<playwright-describe>Interception.setOfflineMode</playwright-describe>
+        ///<playwright-it>should emulate navigator.onLine</playwright-it>
+        [Fact]
```
Comment (kblok) [inline src/PlaywrightSharp.Tests/RequestInterception/InterceptionSetOfflineModeTests.cs]: Skip firefox

File: src/PlaywrightSharp.Tests/RequestInterception/PageSetCacheEnabledTests.cs [hunk 21]
```diff
@@ -0,0 +1,36 @@
+using System.Threading.Tasks;
+using Microsoft.Extensions.Primitives;
+using PlaywrightSharp.Tests.BaseTests;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace PlaywrightSharp.Tests.RequestInterception
```
Comment (kblok) [inline src/PlaywrightSharp.Tests/RequestInterception/PageSetCacheEnabledTests.cs]: Move this test to the Page folder

File: src/PlaywrightSharp.Abstractions/IRequest.cs (hunk 7 above)
Comment (kblok) [reply src/PlaywrightSharp.Abstractions/IRequest.cs]: You missed this one @Meir017 🤓

**Pull request merged**